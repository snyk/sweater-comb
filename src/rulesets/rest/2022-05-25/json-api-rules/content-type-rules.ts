import { RuleError, ResponseRule } from "@useoptic/rulesets-base";
import { links } from "../../../../docs";
import { isOpenApiPath } from "../utils";

export const jsonApiContentTypeRule = new ResponseRule({
  name: "JSON:API content type",
  docsLink: links.jsonApi.contentType,
  matches: (response, rulesContext) =>
    !isOpenApiPath(rulesContext.operation.path) &&
    response.statusCode !== "204" &&
    response.statusCode !== "303",
  rule: (responseAssertions) => {
    responseAssertions.added("use the JSON:API content type", (response) => {
      const responseWithJsonApiContentType = response.bodies.find(
        (body) => body.contentType === "application/vnd.api+json",
      );
      if (!responseWithJsonApiContentType) {
        throw new RuleError({
          message: `expected response to support application/vnd.api+json`,
        });
      }
    });
    responseAssertions.changed(
      "use the JSON:API content type",
      (beforeResponse, response) => {
        const responseWithJsonApiContentType = response.bodies.find(
          (body) => body.contentType === "application/vnd.api+json",
        );
        if (!responseWithJsonApiContentType) {
          throw new RuleError({
            message: `expected response to support application/vnd.api+json`,
          });
        }
      },
    );
  },
});
