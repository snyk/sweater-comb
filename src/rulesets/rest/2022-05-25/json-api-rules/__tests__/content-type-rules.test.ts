import { OpenAPIV3 } from "@useoptic/openapi-utilities";
import { RuleRunner, TestHelpers } from "@useoptic/rulesets-base";
import { context } from "../../__tests__/fixtures";
import { jsonApiContentTypeRule } from "../content-type-rules";

const baseJson = TestHelpers.createEmptySpec();

describe("content type rules", () => {
  test("fails when response does not have json:api content type", () => {
    const beforeJson: OpenAPIV3.Document = {
      ...baseJson,
      paths: {
        "/api/example": {
          get: {
            responses: {},
          },
        },
      },
    };

    const afterJson: OpenAPIV3.Document = {
      ...baseJson,
      paths: {
        "/api/example": {
          get: {
            responses: {
              "200": {
                description: "",
                content: {
                  "application/json": {
                    schema: {
                      type: "object",
                      properties: {},
                    },
                  },
                },
              },
            },
          },
        },
      },
    };

    const ruleRunner = new RuleRunner([jsonApiContentTypeRule]);
    const ruleInputs = {
      ...TestHelpers.createRuleInputs(beforeJson, afterJson),
      context,
    };
    const results = ruleRunner.runRulesWithFacts(ruleInputs);
    expect(results.length).toBeGreaterThan(0);
    expect(results.every((result) => result.passed)).toBe(false);
    expect(results).toMatchSnapshot();
  });
});
