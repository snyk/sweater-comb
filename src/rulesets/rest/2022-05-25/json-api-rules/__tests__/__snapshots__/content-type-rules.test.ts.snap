// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`content type rules fails when response does not have json:api content type 1`] = `
[
  {
    "change": {
      "added": {
        "description": "",
        "statusCode": "200",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "inResponse": {
            "statusCode": "200",
          },
          "method": "get",
          "path": "/api/example",
        },
        "conceptualPath": [
          "operations",
          "/api/example",
          "get",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1api~1example/get/responses/200",
        "kind": "response",
      },
    },
    "condition": undefined,
    "docsLink": "https://jsonapi.org/format/#content-negotiation-clients",
    "error": "expected response to support application/vnd.api+json",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "JSON:API content type",
    "passed": false,
    "received": undefined,
    "type": "added",
    "where": "GET /api/example response 200",
  },
]
`;
